chainerx_cc/chainerx/routines/manipulation.cc

```diff
@@ -555,32 +508,38 @@ Array Stack(const std::vector<Array>& arrays, int8_t axis) {
         if (shape != array.shape()) {
             throw DimensionError{"All input arrays must have the same shape"};
         }
-        // TODO(imanishi): dtype conversion
-        CheckEqual(dtype, array.dtype());
     }
     shape.insert(shape.begin() + axis, static_cast<int64_t>(arrays.size()));
 
-    Strides strides{shape, dtype};
+    Strides strides{shape, out_dtype};
 
     // Aligning with NumPy strides behavior
     auto last_zero_it = std::find(shape.rbegin(), shape.rend(), int64_t{0});
     if (last_zero_it != shape.rend()) {
         std::fill(strides.rbegin() + (last_zero_it - shape.rbegin() + 1), strides.rend(), int64_t{0});
     }
 
-    Array out = internal::Empty(shape, dtype, strides, device);
+    Array out = internal::Empty(shape, out_dtype, strides, device);
+
+    size_t in_size = arrays.size();
+
+    // If input dtypes are mixed, elements in the input arrays are casted to the resulting dtype.
+    // Their original dtypes must therefore be remembered in order to cast the computed gradients back in the backward pass.
+    std::vector<Dtype> in_dtypes;
+    in_dtypes.reserve(in_size);
 
-    if (out.GetTotalSize() != 0) {
-        int64_t step = strides[axis];
-        strides.erase(strides.begin() + axis);
-        {
-            NoBackpropModeScope scope{};
-            int64_t out_offset = 0;
-            for (const Array& array : arrays) {
-                Array sliced_out = internal::MakeArray(array.shape(), strides, dtype, device, out.data(), out_offset);
+    int64_t step = strides[axis];
+    strides.erase(strides.begin() + axis);
+    {
+        NoBackpropModeScope scope{};
+        int64_t out_offset = 0;
+        for (const Array& array : arrays) {
+            if (out.GetTotalSize() != 0) {
```

Comment:
Moved the condition to inside the loop since we need to loop over inputs to collect their dtypes for the backward anyway.